// ember-cli-google-analytics/dist/assets/dummy.js
define('dummy/app', ['exports', 'ember', 'ember/resolver', 'ember/load-initializers', 'dummy/config/environment'], function (exports, Ember, Resolver, loadInitializers, config) {

  'use strict';

  Ember['default'].MODEL_FACTORY_INJECTIONS = true;

  var App = Ember['default'].Application.extend({
    modulePrefix: config['default'].modulePrefix,
    podModulePrefix: config['default'].podModulePrefix,
    Resolver: Resolver['default']
  });

  loadInitializers['default'](App, config['default'].modulePrefix);

  exports['default'] = App;

});
define('dummy/initializers/app-version', ['exports', 'dummy/config/environment', 'ember'], function (exports, config, Ember) {

  'use strict';

  var classify = Ember['default'].String.classify;

  exports['default'] = {
    name: "App Version",
    initialize: function (container, application) {
      var appName = classify(application.toString());
      Ember['default'].libraries.register(appName, config['default'].APP.version);
    }
  };

});
define('dummy/initializers/export-application-global', ['exports', 'ember', 'dummy/config/environment'], function (exports, Ember, config) {

  'use strict';

  exports.initialize = initialize;

  function initialize(container, application) {
    var classifiedName = Ember['default'].String.classify(config['default'].modulePrefix);

    if (config['default'].exportApplicationGlobal) {
      window[classifiedName] = application;
    }
  };

  exports['default'] = {
    name: "export-application-global",

    initialize: initialize
  };

});
define('dummy/mixins/google-pageview', ['exports', 'ember', 'dummy/config/environment'], function (exports, Ember, ENV) {

  'use strict';

  exports['default'] = Ember['default'].Mixin.create({
    pageviewToGA: (function () {
      if (Ember['default'].get(ENV['default'], "googleAnalytics.webPropertyId") != null) {
        var trackerType = Ember['default'].getWithDefault(ENV['default'], "googleAnalytics.tracker", "analytics.js");

        if (trackerType === "analytics.js") {
          var globalVariable = Ember['default'].getWithDefault(ENV['default'], "googleAnalytics.globalVariable", "ga");

          window[globalVariable]("send", "pageview", {
            page: this.get("url"),
            title: this.get("url")
          });
        } else if (trackerType === "ga.js") {
          window._gaq.push(["_trackPageview"]);
        }
      }
    }).on("didTransition")
  });

});
define('dummy/router', ['exports', 'ember', 'dummy/config/environment'], function (exports, Ember, config) {

  'use strict';

  var Router = Ember['default'].Router.extend({
    location: config['default'].locationType
  });

  Router.map(function () {});

  exports['default'] = Router;

});
define('dummy/templates/application', ['exports', 'ember'], function (exports, Ember) {

  'use strict';

  exports['default'] = Ember['default'].Handlebars.template(function anonymous(Handlebars,depth0,helpers,partials,data) {
  this.compilerInfo = [4,'>= 1.0.0'];
  helpers = this.merge(helpers, Ember['default'].Handlebars.helpers); data = data || {};
    var buffer = '', stack1;


    data.buffer.push("<h2 id=\"title\">Welcome to Ember.js</h2>\n\n");
    stack1 = helpers._triageMustache.call(depth0, "outlet", {hash:{},hashTypes:{},hashContexts:{},contexts:[depth0],types:["ID"],data:data});
    if(stack1 || stack1 === 0) { data.buffer.push(stack1); }
    data.buffer.push("\n");
    return buffer;
    
  });

});
define('dummy/tests/app.jshint', function () {

  'use strict';

  module('JSHint - .');
  test('app.js should pass jshint', function() { 
    ok(true, 'app.js should pass jshint.'); 
  });

});
define('dummy/tests/dummy/app.jscs-test', function () {

  'use strict';

  module("JSCS - dummy");
  test("dummy/app.js should pass jscs", function () {
    ok(true, "dummy/app.js should pass jscs.");
  });

});
define('dummy/tests/dummy/initializers/app-version.jscs-test', function () {

  'use strict';

  module("JSCS - dummy/initializers");
  test("dummy/initializers/app-version.js should pass jscs", function () {
    ok(true, "dummy/initializers/app-version.js should pass jscs.");
  });

});
define('dummy/tests/dummy/initializers/export-application-global.jscs-test', function () {

  'use strict';

  module("JSCS - dummy/initializers");
  test("dummy/initializers/export-application-global.js should pass jscs", function () {
    ok(true, "dummy/initializers/export-application-global.js should pass jscs.");
  });

});
define('dummy/tests/dummy/mixins/google-pageview.jscs-test', function () {

  'use strict';

  module("JSCS - dummy/mixins");
  test("dummy/mixins/google-pageview.js should pass jscs", function () {
    ok(true, "dummy/mixins/google-pageview.js should pass jscs.");
  });

});
define('dummy/tests/dummy/router.jscs-test', function () {

  'use strict';

  module("JSCS - dummy");
  test("dummy/router.js should pass jscs", function () {
    ok(true, "dummy/router.js should pass jscs.");
  });

});
define('dummy/tests/dummy/tests/helpers/resolver.jscs-test', function () {

  'use strict';

  module("JSCS - dummy/tests/helpers");
  test("dummy/tests/helpers/resolver.js should pass jscs", function () {
    ok(true, "dummy/tests/helpers/resolver.js should pass jscs.");
  });

});
define('dummy/tests/dummy/tests/helpers/start-app.jscs-test', function () {

  'use strict';

  module("JSCS - dummy/tests/helpers");
  test("dummy/tests/helpers/start-app.js should pass jscs", function () {
    ok(true, "dummy/tests/helpers/start-app.js should pass jscs.");
  });

});
define('dummy/tests/dummy/tests/test-helper.jscs-test', function () {

  'use strict';

  module("JSCS - dummy/tests");
  test("dummy/tests/test-helper.js should pass jscs", function () {
    ok(true, "dummy/tests/test-helper.js should pass jscs.");
  });

});
define('dummy/tests/helpers/resolver', ['exports', 'ember/resolver', 'dummy/config/environment'], function (exports, Resolver, config) {

  'use strict';

  var resolver = Resolver['default'].create();

  resolver.namespace = {
    modulePrefix: config['default'].modulePrefix,
    podModulePrefix: config['default'].podModulePrefix
  };

  exports['default'] = resolver;

});
define('dummy/tests/helpers/resolver.jshint', function () {

  'use strict';

  module('JSHint - helpers');
  test('helpers/resolver.js should pass jshint', function() { 
    ok(true, 'helpers/resolver.js should pass jshint.'); 
  });

});
define('dummy/tests/helpers/start-app', ['exports', 'ember', 'dummy/app', 'dummy/router', 'dummy/config/environment'], function (exports, Ember, Application, Router, config) {

  'use strict';



  exports['default'] = startApp;
  function startApp(attrs) {
    var application;

    var attributes = Ember['default'].merge({}, config['default'].APP);
    attributes = Ember['default'].merge(attributes, attrs); // use defaults, but you can override;

    Ember['default'].run(function () {
      application = Application['default'].create(attributes);
      application.setupForTesting();
      application.injectTestHelpers();
    });

    return application;
  }

});
define('dummy/tests/helpers/start-app.jshint', function () {

  'use strict';

  module('JSHint - helpers');
  test('helpers/start-app.js should pass jshint', function() { 
    ok(true, 'helpers/start-app.js should pass jshint.'); 
  });

});
define('dummy/tests/router.jshint', function () {

  'use strict';

  module('JSHint - .');
  test('router.js should pass jshint', function() { 
    ok(true, 'router.js should pass jshint.'); 
  });

});
define('dummy/tests/test-helper', ['dummy/tests/helpers/resolver', 'ember-qunit'], function (resolver, ember_qunit) {

	'use strict';

	ember_qunit.setResolver(resolver['default']);

});
define('dummy/tests/test-helper.jshint', function () {

  'use strict';

  module('JSHint - .');
  test('test-helper.js should pass jshint', function() { 
    ok(true, 'test-helper.js should pass jshint.'); 
  });

});
/* jshint ignore:start */

define('dummy/config/environment', ['ember'], function(Ember) {
  var prefix = 'dummy';
/* jshint ignore:start */

try {
  var metaName = prefix + '/config/environment';
  var rawConfig = Ember['default'].$('meta[name="' + metaName + '"]').attr('content');
  var config = JSON.parse(unescape(rawConfig));

  return { 'default': config };
}
catch(err) {
  throw new Error('Could not read config from meta tag with name "' + metaName + '".');
}

/* jshint ignore:end */

});

if (runningTests) {
  require("dummy/tests/test-helper");
} else {
  require("dummy/app")["default"].create({"name":"ember-cli-google-analytics","version":"1.3.1.37026be6"});
}

/* jshint ignore:end */
